src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -64126,6 +64126,34 @@ void local2()
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F").WithLocation(18, 13));
         }
 
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71419")]
+        public void LocalFunction_Member_SetInOneFunction()
+        {
+            var source = """
+                #nullable enable
+                class C
```

Comment:
>class C [](http://example.com/codeflow?start=16&length=7)

Before the fix, did we get expected  behavior when this was a `struct` rather than a `class`? Consider adding a test for this scenario as well.

Reply:
>> Before the fix, did we get expected behavior when this was a struct rather than a class? Consider adding a test for this scenario as well.

> Yes, we got expected behavior. ...

This tells me that there is likely something different between fields (accessed through captured `this`) that do not get special treatment when we create the special state, and regular captured locals that get special treatment when we create the state. Is there a similar problem with regular captured locals? If the answer is "no", then it feels to me, that flipping between reachable and unreachable aspect isn't fixing the root cause of the problem, even though it eliminates the "symptom" for the specific scenario. The reason for that feeling is an assumption that reachable/unreachable aspect has nothing to do with unformal handling of locals and fields. If this assumption is wrong, then I would like to understand why.

Reply:
The difference is that the nullable walker tracks classes differently than structs.
- For class, the state is initially just `this!` - the field is not tracked since we haven't encountered its access yet. Hence the initial `localFunctionState.StartingState` is also `this!`. Then at `local1();` call, the current state is `this!this.F!`, the `localFunctionState.StartingState` is normalized to `this!this.F?` and joined with the current state, hence the field is incorrectly marked maybe-null (the bug is fixed by making the `localFunctionState.StartingState` unreachable initially so it's discarded during this join).
- For struct, the state is initially already `this!this.F!`. Hence the initial `localFunctionState.StartingState` is also `this!this.F` and the bug doesn't surface.

I don't think the root cause of the problem is how structs/classes are handled by the nullable walker. I think the root cause is that the `localFunctionState.StartingState` (which can be described as the worst state at all use-sites of the local function) is initially (when we haven't seen any use-site yet) reachable. If it's instead unreachable initially, it will be discarded and overwritten when we first encounter a use-site of the local function. The initial value will only be preserved if the local function is never used.

Reply:
> It looks obvious to me that we do not treat all tracked "variables" uniformly

I agree and I have arrived at the same conclusion, i.e., we can make `ReachableStateWithNotNulls` remember that it should set newly seen variables as "not null". I can investigate this route, then open a separate PR.

However, I still believe the change in this PR is correct in any case - the initial local function use-site state is simply unreachable, so it should be marked as such. (Plus it fixes a regression which would be good to fix as soon as possible; although that could be arguably also solved by reverting the PR that caused it.)

Reply:
> I can investigate this route, then open a separate PR.

I have prototyped this approach: https://github.com/dotnet/roslyn/pull/71653.

However, I'm not sure how to fix the example from https://github.com/dotnet/roslyn/pull/71506#issuecomment-1889261840 (test `LocalFunction_Member_SetInOneFunction_Loop`) with this approach. On the other hand, it seems simple with the approach in the current PR: https://github.com/dotnet/roslyn/pull/71654/commits/b235ace82e78165260d8dbc3e27704d9d74b69f9 (analyze the local function body as unreachable - hence the loop head state is also marked as unreachable and overwritten later by a reachable state)

Reply:
Here is an example that, in my opinion, demonstrates the fact that an allocation of a slot prior to visiting the local function makes a difference:
```
class C
{
  string? F;

  void M2()
  {
    F = null; // We allocate slot here
    local1();

    void local2()
    {
      F.ToString(); // No more CS8602: Dereference of a possibly null reference.
    }

    void local1()
    {
      F = "";
      local2();
    }
  }
}
```

> However, I still believe the change in this PR is correct in any case - the initial local function use-site state is simply unreachable, so it should be marked as such.

Why bother visiting the local function then? What benefit do we get from doing all the work? Do we have an explanation?

> Plus it fixes a regression which would be good to fix as soon as possible

Just the fact that the specific scenario appears to be fixed by a change, doesn't mean that the change is the right change to make. It has been observed on several occasions that two bugs can lead to an expected behavior for some specific scenario. Bugs could "mask" each other for some specific scenario, making problems unobservable for that scenario. That, however, doesn't change the fact that they are bugs. Given that, in such non-obvious case, the fact that the proposed change appears to change the observable behavior to the better, isn't sufficient to confirm its correctness for me. Having reachable/unreachable state wrong feels like very fundamental problem to me. Therefore, I would expect it to manifest itself in a broader set of scenarios and have broader impact. Basically, in a case like this, I would look for a confirmation of the theory from several different "sources". For example, finding scenarios of a different nature also negatively affected by this and confirming they are fixed as well. 
The situation around ```ReachableStateWithNotNulls``` looks more obvious to me, because the intent of the function is documented, the fact that the implementation doesn't fulfil the intent can be confirmed by examining the implementation, the fact that allocating a slot before the analysis changes things to the better can be observed in the example at the beginning of this message. 


>  I'm not sure how to fix the example from https://github.com/dotnet/roslyn/pull/71506#issuecomment-1889261840 (test `LocalFunction_Member_SetInOneFunction_Loop`) with this approach.

It is quite possible I am missing something. If what we are doing today for locals is sufficient for correct behavior, then, after we align behavior for other kinds of "variables" with behavior for locals, we should get the same correct behavior for them  

> On the other hand, it seems simple with the approach in the current PR

I think that simplicity isn't a factor that can help us to choose what road to take, when the roads lead to different destinations. It could help us make a decision in a case when we arrive to the same point, but the alternatives that we are discussing do not lead us to the same point.

Reply:
I think the general problem that keeps coming up is that maybe-null state for a given variable from the "default starting state" for local functions keeps leaking into the subsequent analysis passes on the local function. It may leak into the "main" State property or into state fields which are preserved across analysis passes such as `loopHeadState`.

I think previously, we wanted to solve this problem by giving captured variables not-null initial state. But I don't think this fully solves the problem. Not just because our fix failed to give a not-null state for nested slots in some cases, but because I think we get a better development experience when doing the "default starting state" analysis pass with certain variables having maybe-null state, such as properties/fields where the declared type is nullable. So I don't think it works to just make everything not-null, in order to avoid having maybe-nulls leak into the subsequent passes.

One way we could solve this problem would be to simply *only* do the "default" visit when the local function is not used anywhere. When the local function *is* used, we can skip that visit and prevent the "default starting state" from leaking in to the subsequent passes, and also save a little bit of unnecessary work, by only analyzing the local function body based on the call sites.

Reply:
> One way we could solve this problem, which I think Aleksey alluded to, would be to simply _only_ do the "default" visit when the local function is not used anywhere. When the local function _is_ used, we can skip that visit and prevent the "default starting state" from leaking in to the subsequent passes, and also save a little bit of unnecessary work, by only analyzing the local function body based on the call sites.

Great, if you all agree that's a good solution, I'm in (I already proposed it in https://github.com/dotnet/roslyn/pull/71506#issuecomment-1889467221 - at least I hope it's equivalent to what you are suggesting). I have started a PR implementing this in https://github.com/dotnet/roslyn/pull/71679.